### PR TITLE
feat(vendor): support modifying remote files in vendor folder without checksum errors

### PR DESCRIPTION
**Pull request by dsherret**
Includes:

* https://github.com/denoland/deno_graph/pull/486
* https://github.com/denoland/deno_graph/pull/488
* https://github.com/denoland/deno_lockfile/pull/25
* https://github.com/denoland/deno_lockfile/pull/22
* https://github.com/denoland/deno_graph/pull/483
* https://github.com/denoland/deno_graph/pull/470

This is actually a lockfile perf investigation of moving the lockfile verification to be done in deno_graph's loader, but it has the side effect of enabling the ability to modify remote modules in the `vendor` folder without causing checksum errors.

**Benchmarks**

```
% git clone https://github.com/jsr-io/jsr
% git switch deno_graph_slow
% cd frontend
% hyperfine --warmup 3 "../../deno/deno_old run -A repro.ts" "../../deno/target/release/deno run -A repro.ts"
```

With `"vendor": true` (note: this PR doesn't fix the underlying issue that makes this scenario slow... I will follow up with that later):

```
Benchmark 1: ../../deno/deno_old run -A repro.ts
  Time (mean ± σ):     538.0 ms ±   5.5 ms    [User: 444.0 ms, System: 101.5 ms]
  Range (min … max):   531.9 ms … 548.6 ms    10 runs
 
Benchmark 2: ../../deno/target/release/deno run -A repro.ts
  Time (mean ± σ):     458.5 ms ±   3.1 ms    [User: 366.4 ms, System: 100.5 ms]
  Range (min … max):   453.4 ms … 463.7 ms    10 runs
 
Summary
  ../../deno/target/release/deno run -A repro.ts ran
    1.17 ± 0.01 times faster than ../../deno/deno_old run -A repro.ts
```

With `"vendor": false` (slightly slower because it now has to verify checksums):

```
Benchmark 1: ../../deno/deno_old run -A repro.ts
  Time (mean ± σ):     539.5 ms ±   3.7 ms    [User: 447.8 ms, System: 99.8 ms]
  Range (min … max):   536.2 ms … 546.0 ms    10 runs
 
Benchmark 2: ../../deno/target/release/deno run -A repro.ts
  Time (mean ± σ):     463.6 ms ±   8.3 ms    [User: 371.3 ms, System: 100.8 ms]
  Range (min … max):   456.0 ms … 484.2 ms    10 runs
```

### CODE REVIEW

File: tests/integration/jsr_tests.rs [hunk 1]
```diff
@@ -341,33 +347,33 @@ console.log(add);"#,
...
   test_context
     .new_command()
-    .args("run  main.ts")
+    .args("run main.ts")
     .run()
     .assert_matches_text(
-      // ideally the two error messages would be the same... this one comes from
-      // deno_cache and the one above comes from deno_graph. The thing is, in deno_cache
-      // (source of this error) it makes sense to include the url in the error message
-      // because it's not always used in the context of deno_graph
```
Comment (dsherret) [inline tests/integration/jsr_tests.rs:362]: Fixed!

File: cli/module_loader.rs [hunk 2]
```diff
@@ -173,13 +173,9 @@ impl ModuleLoadPreparer {
 
     self.module_graph_builder.graph_roots_valid(graph, roots)?;
 
-    // If there is a lockfile...
+    // write the lockfile if there is one
     if let Some(lockfile) = &self.lockfile {
-      let mut lockfile = lockfile.lock();
-      // validate the integrity of all the modules
-      graph_lock_or_exit(graph, &mut lockfile);
```
Comment (dsherret) [inline cli/module_loader.rs:180]: Handled in graph_roots_valid.

File: cli/graph_util.rs [hunk 3]
```diff
@@ -578,6 +619,53 @@ impl ModuleGraphBuilder {
...
+      graph.packages.package_deps_sum() != initial_package_deps_len;
+    let has_jsr_package_mappings_changed =
+      graph.packages.mappings().len() != initial_package_mappings_len;
+    let has_npm_packages_changed =
+      graph.npm_packages.len() != initial_npm_packages;
+
+    if has_redirects_changed
+      || has_jsr_package_deps_changed
+      || has_jsr_package_mappings_changed
+      || has_npm_packages_changed
```
Comment (dsherret) [inline cli/graph_util.rs:633]: Todo: I'll add a test for each of these individual conditions before landing this PR.

File: cli/Cargo.toml [hunk 4]
```diff
@@ -67,17 +67,17 @@ deno_ast = { workspace = true, features = ["bundler", "cjs", "codegen", "proposa
 deno_cache_dir = { workspace = true }
 deno_config = "=0.16.3"
 deno_core = { workspace = true, features = ["include_js_files_for_snapshotting"] }
-deno_doc = { version = "=0.135.0", features = ["html", "syntect"] }
-deno_emit = "=0.40.3"
-deno_graph = { version = "=0.75.2", features = ["tokio_executor"] }
+deno_doc = { version = "=0.137.0", features = ["html", "syntect"] }
+deno_emit = "=0.41.0"
+deno_graph = { version = "=0.77.0", features = ["tokio_executor"] }
```
Comment (dsherret) [inline cli/Cargo.toml:72]: It seems some wpt tests are failing due to https://github.com/denoland/deno_graph/pull/483

See https://github.com/denoland/deno/actions/runs/9262872207

I'm unsure how to interpret the wpt test output. Are we incorrectly failing now?

Comment (lucacasonato) [reply cli/Cargo.toml:72]: I have updated the expectations. We were previously failing these tests, now the tests pass. The expectations file was assuming the tests to fail though, so them passing caused the suite to fail.

**Review by marvinhagemeister — APPROVED** [review]
LGTM, went through the PR 👍

**Pull request merged**